lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -502,26 +509,17 @@ public void OnNext(IFailedEvaluator failedEvaluator)
                     switch (_systemState.CurrentState)
                     {
                         case SystemState.WaitingForEvaluator:
-                            if (!_evaluatorManager.ExceededMaximumNumberOfEvaluatorFailures())
+                            if (!_evaluatorManager.ExceededMaximumNumberOfEvaluatorFailures() && !isMaster)
                             {
-                                if (isMaster)
-                                {
-                                    Logger.Log(Level.Info, "Requesting a master Evaluator.");
-                                    _evaluatorManager.RemoveFailedEvaluator(failedEvaluator.Id);
-                                    _evaluatorManager.RequestUpdateEvaluator();
-                                }
-                                else
-                                {
-                                    _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(
-                                        failedEvaluator.Id);
-                                    Logger.Log(Level.Info, "Requesting mapper Evaluators.");
-                                    _evaluatorManager.RemoveFailedEvaluator(failedEvaluator.Id);
-                                    _evaluatorManager.RequestMapEvaluators(1);
-                                }
+                                _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(
+                                    failedEvaluator.Id);
+                                Logger.Log(Level.Info, "Requesting mapper Evaluators.");
+                                _evaluatorManager.RemoveFailedEvaluator(failedEvaluator.Id);
+                                _evaluatorManager.RequestMapEvaluators(1);
                             }
                             else
                             {
-                                Logger.Log(Level.Error, "The system is not recoverable, change the state to Fail.");
+                                Logger.Log(Level.Error, "The system is not recoverable, either because it exceeded MaximumNumberOfEvaluatorFailures or master evaluator failed, change the state to Fail.");
```

Comment:
When we debug job failure, we'll need to know which cause it was. If we're logging more detailed reason for failure later, we don't need this either-or here; if this is the only log we get, we'd better log a specific reason here